### DISCUSSION TITLE
Is there any way to monitor all jobs

### DISCUSSION THREAD

**Discussion by iakirago** [go-co-op/gocron#513] [Q&A] [answered]
Like python's apscheduler, it can complete the common processing of all tasks in the monitor

**Comment by JohnRoesler** [comment]
I'm not familiar with apscheduler. What information would you expect to see when monitoring? Is that an application call, a cli, or some sort of UI?

**Reply by iakirago** [reply]
I hope it is an application callback. When adding a callback, I can specify that the callback method will be executed when the task execution is completed or an exception occurs.

**Reply by JohnRoesler** [reply]
For execution completion you can use: https://pkg.go.dev/github.com/go-co-op/gocron#Job.SetEventListeners

there is an open ticket to handle the on error scenario https://github.com/go-co-op/gocron/issues/468

**Reply by iakirago** [reply]
I read the sample code, but found that this is only an event listener added for a single job, not an event listener for s in `s := gocron.NewScheduler(time.UTC),` I want to have a global event listener that can listen to all jobs

**Reply by JohnRoesler** [reply]
You could use the same functions and pass them to all the jobs today. 

I think that is a reasonable feature to request, the ability to set the event listeners globally at the scheduler to apply for all jobs.

**Reply by iakirago** [reply]
Because I currently have more than 50 jobs, manually adding monitors to each job is a relatively large workload, and there will be a lot of repeated code. For now, I can only manually add listeners to each job. I would like to know if there is a development plan for the global monitoring function, and if there is a corresponding follow-up issue, when will this function be implemented?

**Reply by JohnRoesler** [reply, marked as answer]
Adding listeners to all the jobs is trivial, for example:

```golang
package main

import (
	"fmt"
	"time"

	"github.com/go-co-op/gocron"
)

func myBeforeJobRunsFunc() {
	fmt.Println("myBeforeJobRunsFunc")
}
func myAfterJobRunsFunc() {
	fmt.Println("myAfterJobRunsFunc")
}

func main() {
	s := gocron.NewScheduler(time.UTC)

	s.Every("1s").Do(func() {})
	s.Every("1s").Do(func() {})
	s.Every("1s").Do(func() {})
	s.Every("1s").Do(func() {})
	s.Every("1s").Do(func() {})

	for _, job := range s.Jobs() {
		job.SetEventListeners(myBeforeJobRunsFunc, myAfterJobRunsFunc)
	}
	s.StartBlocking()
}
```

> I would like to know if there is a development plan for the global monitoring function, and if there is a corresponding follow-up issue, when will this function be implemented?

There is not yet a feature request issue created for this. Feel free to open one and describe the request. As for a timeline, as this is an opensource project, I work on it as I have time. You are free to contribute if you'd like to see the feature done more quickly!

**Reply by iakirago** [reply]
Thank you very much for your answer, this solves my problem for now. I have submitted the corresponding feature issue, please help to review it.
